--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
@@ -18,15 +_,39 @@
     private final int z;
     private final ClientboundLevelChunkPacketData chunkData;
     private final ClientboundLightUpdatePacketData lightData;
+    // Spout start - Block mapping - Chunk packets - Call
+    public spout.server.paper.impl.packetmapping.block.chunk.@Nullable ChunkPacketBlockMapper spoutChunkPacketBlockMapper;
+    private final java.util.concurrent.atomic.AtomicBoolean startedMappingBlocks = new java.util.concurrent.atomic.AtomicBoolean();
+    public volatile boolean readyMappingBlocks;
+    // Spout end - Block mapping - Chunk packets - Call
     // Paper start - Async-Anti-Xray - Ready flag for the connection, add chunk packet info
     private volatile boolean ready;
     @Override
     public boolean isReady() {
-        return this.ready;
+        // Spout start - Block mapping - Chunk packets - Call after x-ray
+        if (!this.ready) {
+            return false;
+        }
+        if (!this.readyMappingBlocks) {
+            this.startMappingBlocksIfNecessary();
+        }
+        return this.readyMappingBlocks;
+    }
+
+    private void startMappingBlocksIfNecessary() {
+        if (!this.startedMappingBlocks.compareAndExchange(false, true)) {
+            if (this.spoutChunkPacketBlockMapper != null) {
+                this.spoutChunkPacketBlockMapper.applyMappings();
+            } else {
+                this.readyMappingBlocks = true;
+            }
+        }
     }
 
     public void setReady(final boolean ready) {
-        this.ready = ready;
+        if (!this.ready) {
+            this.ready = ready;
+            if (!org.bukkit.Bukkit.isPrimaryThread()) {
+                this.startMappingBlocksIfNecessary();
+            }
+        }
+        // Spout end - Block mapping - Chunk packets - Call after x-ray
     }
     // Paper end - Async-Anti-Xray - Ready flag for the connection, add chunk packet info
 
@@ -36,9 +_,10 @@
         final LevelLightEngine lightEngine,
         final @Nullable BitSet skyChangedLightSectionFilter,
         final @Nullable BitSet blockChangedLightSectionFilter
+        , final net.minecraft.server.level.ServerPlayer player // Spout start - Block mapping - Chunk packets - Pass the player
     ) {
         // Paper start - Anti-Xray
-        this(levelChunk, lightEngine, skyChangedLightSectionFilter, blockChangedLightSectionFilter, true);
+        this(levelChunk, lightEngine, skyChangedLightSectionFilter, blockChangedLightSectionFilter, true, player); // Spout start - Block mapping - Chunk packets - Pass the player
     }
     public ClientboundLevelChunkWithLightPacket(
         final LevelChunk levelChunk,
@@ -46,15 +_,17 @@
         final @Nullable BitSet skyChangedLightSectionFilter,
         final @Nullable BitSet blockChangedLightSectionFilter,
         final boolean modifyBlocks
+        , final net.minecraft.server.level.ServerPlayer player // Spout start - Block mapping - Chunk packets - Pass the player
     ) {
         // Paper end - Anti-Xray
         ChunkPos chunkPos = levelChunk.getPos();
         this.x = chunkPos.x();
         this.z = chunkPos.z();
         io.papermc.paper.antixray.ChunkPacketInfo<net.minecraft.world.level.block.state.BlockState> chunkPacketInfo = modifyBlocks ? levelChunk.getLevel().chunkPacketBlockController.getChunkPacketInfo(this, levelChunk) : null; // Paper - Anti-Xray
+        if (chunkPacketInfo == null) chunkPacketInfo = new io.papermc.paper.antixray.ChunkPacketInfo<>(this, levelChunk); // Spout - Block mapping - Chunk packets - Empty chunks still need section indexes
         this.chunkData = new ClientboundLevelChunkPacketData(levelChunk, chunkPacketInfo); // Paper - Anti-Xray
         this.lightData = new ClientboundLightUpdatePacketData(chunkPos, lightEngine, skyChangedLightSectionFilter, blockChangedLightSectionFilter);
         levelChunk.getLevel().chunkPacketBlockController.modifyBlocks(this, chunkPacketInfo); // Paper - Anti-Xray - Modify blocks
+        this.spoutChunkPacketBlockMapper = new spout.server.paper.impl.packetmapping.block.chunk.ChunkPacketBlockMapper(this, levelChunk, player, chunkPacketInfo);
     }
 
     private ClientboundLevelChunkWithLightPacket(final RegistryFriendlyByteBuf input) {
