--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -175,23 +_,72 @@
         return new StreamCodec<RegistryFriendlyByteBuf, ItemStack>() {
             @Override
             public ItemStack decode(final RegistryFriendlyByteBuf input) {
+                // Spout start - Item mapping - Reverse - Apply at ItemStack decoding
+                return this.decode(input, true);
+            }
+
+            public ItemStack decode(RegistryFriendlyByteBuf input, boolean attemptReversing) {
+                // Spout end - Item mapping - Reverse - Apply at ItemStack decoding
                 int count = input.readVarInt();
                 if (count <= 0) {
                     return ItemStack.EMPTY;
                 } else {
                     Holder<Item> item = Item.STREAM_CODEC.decode(input);
                     DataComponentPatch patch = patchCodec.decode(input);
-                    return new ItemStack(item, count, patch);
+                    // Spout start - Item mapping - Reverse - Apply at ItemStack decoding
+                    ItemStack received = new ItemStack(item, count, patch);
+                    if (attemptReversing) {
+                        spout.server.paper.impl.packetmapping.item.reverse.@Nullable ItemMappingReverser itemMappingReverser = ((spout.server.paper.impl.clientview.ClientViewImpl) input.getClientViewOrFallback()).getItemMappingReverser();
+                        if (itemMappingReverser != null) {
+                            received = itemMappingReverser.reverseMappingIfPossible(received);
+                        }
+                    }
+                    return received;
+                    // Spout end - Item mapping - Reverse - Apply at ItemStack decoding
                 }
             }
 
             @Override
             public void encode(final RegistryFriendlyByteBuf output, final ItemStack itemStack) {
+                // Spout start - Item mapping - Implementation part - Apply at ItemStack encoding
+                spout.server.paper.impl.packetmapping.item.ItemMappingFunctionContextImpl itemMappingContext = new spout.server.paper.impl.packetmapping.item.ItemMappingFunctionContextImpl(output.getClientViewOrFallback(), output.isEncodingItemFrameItem, output.isEncodingStonecutterRecipeResult);
+                spout.server.paper.impl.packetmapping.item.reverse.@Nullable ItemMappingReverser itemMappingReverser = ((spout.server.paper.impl.clientview.ClientViewImpl) output.getClientViewOrFallback()).getItemMappingReverser();
+                ItemStack mappedValue = spout.server.paper.impl.packetmapping.item.ItemMappingsImpl.get().apply(itemStack, itemMappingContext);
+                if (itemMappingReverser != null) {
+                    long reservedId = -1;
+                    if (mappedValue != itemStack) {
+                        reservedId = itemMappingReverser.reserveAndAttachId(mappedValue, itemStack);
+                    }
+                    int writerIndexBeforeEncoding = output.writerIndex();
+                    this.encodeUnmapped(output, mappedValue);
+                    int readerIndexBeforeReading = output.readerIndex();
+                    output.readerIndex(writerIndexBeforeEncoding);
+                    ItemStack actuallySent = this.decode(output, false);
+                    output.readerIndex(readerIndexBeforeReading);
+                    if (!matches(actuallySent, mappedValue)) {
+                        if (reservedId == -1) {
+                            mappedValue = mappedValue.copy();
+                        }
+                        reservedId = itemMappingReverser.reserveAndAttachId(actuallySent, itemStack);
+                        itemMappingReverser.attachId(mappedValue, reservedId);
+                        output.writerIndex(writerIndexBeforeEncoding);
+                        this.encodeUnmapped(output, mappedValue);
+                    }
+                    if (reservedId != -1) {
+                        itemMappingReverser.storeAsReversible(reservedId, actuallySent);
+                    }
+                } else {
+                    this.encodeUnmapped(output, mappedValue);
+                }
+            }
+
+            private void encodeUnmapped(RegistryFriendlyByteBuf output, ItemStack itemStack) {
+                // Spout end - Item mapping - Implementation part - Apply at ItemStack encoding
                 if (itemStack.isEmpty() || itemStack.getItem() == null) { // CraftBukkit - NPE fix itemstack.getItem()
                     output.writeVarInt(0);
                 } else {
                     output.writeVarInt(io.papermc.paper.util.sanitizer.ItemComponentSanitizer.sanitizeCount(io.papermc.paper.util.sanitizer.ItemObfuscationSession.currentSession(), itemStack, itemStack.getCount())); // Paper - potentially sanitize count
-                    Item.STREAM_CODEC.encode(output, itemStack.typeHolder());
+                    Item.UNMAPPED_STREAM_CODEC.encode(output, itemStack.typeHolder()); // Spout - Item mapping - Wrap codec for items - Use original for unmapped item stacks
                     // Paper start - adventure; conditionally render translatable components
                     boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
                     try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(itemStack))) { // pass the itemstack as context to the obfuscation session
@@ -475,6 +_,11 @@
                         if (!(block.getBlock() instanceof net.minecraft.world.level.block.BaseEntityBlock)) { // Containers get placed automatically
                             block.onPlace(level, newPos, oldBlock, true, context);
                         }
+                        // Spout start - More data-driven - Blocks - Refresh light after captured custom block placement
+                        if (net.minecraft.world.level.lighting.LightEngine.hasDifferentLightProperties(oldBlock, block)) {
+                            level.getChunkSource().getLightEngine().checkBlock(newPos);
+                        }
+                        // Spout end - More data-driven - Blocks - Refresh light after captured custom block placement
 
                         level.notifyAndUpdatePhysics(newPos, null, oldBlock, block, level.getBlockState(newPos), updateFlags, net.minecraft.world.level.block.Block.UPDATE_LIMIT); // send null chunk as chunk.k() returns false by this point
                     }
